## 1.0 IMPORT LIBRARIES

In [7]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score ,f1_score,classification_report, make_scorer
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import plot_tree
from sklearn.tree import export_text
from sklearn.utils import resample
from sklearn import preprocessing
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier
from sklearn.linear_model import LogisticRegression
import os
import tensorflow as tf
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
from __future__ import print_function
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn import datasets

# 2.0 LOAD THE DATA

In [8]:
X_train = pd.read_csv('FinancialData_train_X.csv')
X_test = pd.read_csv('FinancialData_test_X.csv')
y_train = pd.read_csv('FinancialData_train_y.csv')
y_test = pd.read_csv('FinancialData_test_y.csv')

In [4]:
X_train.head(5)

ROA(C) before interest and depreciation before interest   
0                                          -7.064528         \
1                                           0.398895          
2                                           1.066762          
3                                          -0.502727          
4                                          -0.152096          

    Operating Gross Margin   Operating Profit Rate   
0                -0.382990              -24.732582  \
1                 0.172025                0.436047   
2                 1.628940                0.436047   
3                -0.313613               -0.531977   
4                -0.729875               -0.531977   

    Pre-tax net Interest Rate   Operating Expense Rate   
0                   -9.954255                -0.623336  \
1                    0.039083                -0.623336   
2                    0.039083                -0.623336   
3                    0.039083                -0.623336   
4                    0.039083                 1.155986   

    Research and development expense rate   Cash flow rate   
0                               -0.750627        -3.421733  \
1                                1.316196         0.185156   
2                               -0.750627         0.735359   
3                               -0.750627        -0.487315   
4                                0.516259        -0.487315   

    Interest-bearing debt interest rate   Tax rate (A)   
0                             -0.146024      -0.830060  \
1                             -0.146024       0.985223   
2                             -0.146024      -0.408400   
3                             -0.146024       3.265046   
4                             -0.146024      -0.830060   

    Net Value Per Share (B)  ...   Current Liability to Current Assets   
0                 -1.679873  ...                             -0.303571  \
1                 -0.089657  ...                              0.236438   
2                  0.137517  ...                             -0.573576   
3                 -0.998352  ...                              3.814000   
4                  0.267331  ...                             -0.168569   

    Liability-Assets Flag   Total assets to GNP price   No-credit Interval   
0               -0.032383                    4.692638            -0.030519  \
1               -0.032383                   -0.053114            -0.030519   
2               -0.032383                   -0.053114            -0.030519   
3               -0.032383                   -0.053114            -0.030519   
4               -0.032383                   -0.053114             0.040261   

    Net Income to Stockholder's Equity   Liability to Equity   
0                            -1.809277             -0.120046  \
1                             0.070906             -0.185017   
2                             0.205204             -0.249988   
3                             0.003756             -0.120046   
4                             0.003756              0.009896   

    Degree of Financial Leverage (DFL)   
0                            -0.052065  \
1                            -0.052065   
2                            -0.052065   
3                             0.055934   
4                             0.001935   

    Interest Coverage Ratio (Interest expense to EBIT)   Net Income Flag   
0                                           0.013626                 0.0  \
1                                           0.013626                 0.0   
2                                           0.013626                 0.0   
3                                           0.157200                 0.0   
4                                           0.085413                 0.0   

    Equity to Liability  
0             -0.201885  
1             -0.096144  
2              0.453711  
3             -0.244182  
4             -0.434516  

[5 rows x 22 columns]

In [5]:
X_train.shape

(4773, 22)

In [6]:
X_train.columns

Index([' ROA(C) before interest and depreciation before interest',
       ' Operating Gross Margin', ' Operating Profit Rate',
       ' Pre-tax net Interest Rate', ' Operating Expense Rate',
       ' Research and development expense rate', ' Cash flow rate',
       ' Interest-bearing debt interest rate', ' Tax rate (A)',
       ' Net Value Per Share (B)', ' CFO to Assets', ' Cash Flow to Equity',
       ' Current Liability to Current Assets', ' Liability-Assets Flag',
       ' Total assets to GNP price', ' No-credit Interval',
       ' Net Income to Stockholder's Equity', ' Liability to Equity',
       ' Degree of Financial Leverage (DFL)',
       ' Interest Coverage Ratio (Interest expense to EBIT)',
       ' Net Income Flag', ' Equity to Liability'],
      dtype='object')

In [7]:
y_train.columns

Index(['Bankrupt'], dtype='object')

# 3.0 Creating dataframe

In [33]:
performance = pd.DataFrame({"model": [], "Accuracy": [], "Precision": [], "Recall": [], "F1": []})

# 4.0 Execution of Models

## Logistic Regression using Random Search CV

In [9]:
score_measure = "accuracy"
kfolds = 5

parameter_grid = {
    'max_iter':np.arange(500,1000),
    'penalty': ['None','l1','l2','elasticnet'],
    'solver':['saga','liblinear']
}

log_reg = LogisticRegression()
random_search = RandomizedSearchCV(estimator = log_reg, param_distributions=parameter_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1, 
                           return_train_score=True)

_ = random_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {random_search.best_score_}")
print(f"... with parameters: {random_search.best_params_}")\

bestrecall = random_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning

The best accuracy score is 0.9694111319657107
... with parameters: {'solver': 'saga', 'penalty': 'l2', 'max_iter': 645}


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


In [10]:
c_matrix = confusion_matrix(y_test, random_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Logistic RandomRegression", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

##  Logistic Regression Grid Search CV

In [11]:
score_measure = "accuracy"
kfolds = 5
max_iter = random_search.best_params_['max_iter']
penalty = random_search.best_params_['penalty']
solver = random_search.best_params_['solver']

parameter_grid = {
    'max_iter': np.arange(max_iter-100,max_iter+100),  
    'penalty': [penalty],
    'solver': [solver]
}

log_reg = LogisticRegression()
grid_search = GridSearchCV(estimator = log_reg, param_grid=parameter_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestrecall = grid_search.best_estimator_

Fitting 5 folds for each of 200 candidates, totalling 1000 fits


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning

The best accuracy score is 0.9694111319657107
... with parameters: {'max_iter': 545, 'penalty': 'l2', 'solver': 'saga'}


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


In [12]:
c_matrix = confusion_matrix(y_test, grid_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Logistic GridRegression", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## SVM Random Search CV using linear, rbf and poly kernal

In [13]:
from sklearn.svm import SVC
score_measure = "accuracy"
kfolds = 5

parameter_grid = {
    'C': np.arange(1,25),   
    'gamma': ['scale','auto'],
    'kernel':['linear','rbf','poly']
}

svm_model = SVC()
random_search = RandomizedSearchCV(estimator = svm_model, param_distributions=parameter_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1, 
                           return_train_score=True)

_ = random_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {random_search.best_score_}")
print(f"... with parameters: {random_search.best_params_}")


bestrecall = random_search.best_estimator_


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/model_selection/_search.py:305: UserWarning: The total space of parameters 144 is smaller than n_iter=500. Running 144 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was 

Fitting 5 folds for each of 144 candidates, totalling 720 fits


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning

The best accuracy score is 0.97066855455673
... with parameters: {'kernel': 'linear', 'gamma': 'scale', 'C': 1}


In [14]:
c_matrix = confusion_matrix(y_test, random_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"SVM Random", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## SVM Grid Search CV using linear,rbf and poly kernal

In [15]:
score_measure = "accuracy"
kfolds = 5

C = random_search.best_params_['C']
gamma = random_search.best_params_['gamma']
kernel = random_search.best_params_['kernel']

parameter_grid = {
    'C': np.arange(C-2,C+2),  
    'gamma': [gamma],
    'kernel': [kernel]
    
}

svm_model = SVC()
grid_search = GridSearchCV(estimator = svm_model, param_grid=parameter_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestrecall = grid_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning

The best accuracy score is 0.97066855455673
... with parameters: {'C': 1, 'gamma': 'scale', 'kernel': 'linear'}


In [16]:
c_matrix = confusion_matrix(y_test, grid_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"SVM Grid", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## Decision tree by using Random search

In [17]:
score_measure = "accuracy"
kfolds = 5

parameter_grid = {
    'min_samples_split': np.arange(1,200),  
    'min_samples_leaf': np.arange(1,200),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 200), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
random_search = RandomizedSearchCV(estimator = dtree, param_distributions=parameter_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = random_search.fit(X_train, np.ravel(y_train))

print(f"The best {score_measure} score is {random_search.best_score_}")
print(f"... with parameters: {random_search.best_params_}")

bestrecall = random_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits
The best accuracy score is 0.9687828597144017
... with parameters: {'min_samples_split': 135, 'min_samples_leaf': 196, 'min_impurity_decrease': 0.0026, 'max_leaf_nodes': 100, 'max_depth': 28, 'criterion': 'entropy'}


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
5 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
5 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/tree/_classes.py", line 889, in fit
    super().fit(
  File "/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/tree/_classes.py", line 177, in fit

In [18]:
c_matrix = confusion_matrix(y_test, random_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Dtree_random", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

/var/folders/f5/wct_d8p94tn748cf64rswjm00000gn/T/ipykernel_24331/3097715638.py:8: RuntimeWarning: invalid value encountered in long_scalars
  'Precision': [TP/(TP+FP)],


## Decision tree by using Grid search

In [19]:
score_measure = "accuracy"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(30,36),  
    'min_samples_leaf': np.arange(6,12),
    'min_impurity_decrease': np.arange(0.0048, 0.0054, 0.0001),
    'max_leaf_nodes': np.arange(162,168), 
    'max_depth': np.arange(15,21), 
    'criterion': ['entropy'],
}

dtree = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = dtree, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, np.ravel(y_train))

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestrecall = grid_search.best_estimator_

Fitting 5 folds for each of 9072 candidates, totalling 45360 fits
The best accuracy score is 0.9687828597144017
... with parameters: {'criterion': 'entropy', 'max_depth': 15, 'max_leaf_nodes': 162, 'min_impurity_decrease': 0.0048, 'min_samples_leaf': 6, 'min_samples_split': 30}


In [20]:
c_matrix = confusion_matrix(y_test, grid_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Dtree_grid", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

/var/folders/f5/wct_d8p94tn748cf64rswjm00000gn/T/ipykernel_24331/732757500.py:8: RuntimeWarning: invalid value encountered in long_scalars
  'Precision': [TP/(TP+FP)],


## Neural Networks

In [21]:
from sklearn.neural_network import MLPClassifier

In [22]:
%%time

ann = MLPClassifier(hidden_layer_sizes=(60,50,40), solver='adam', max_iter=200)
_ = ann.fit(X_train, y_train)

/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


CPU times: user 24.1 s, sys: 3.05 s, total: 27.2 s
Wall time: 3.61 s


In [23]:
%%time
y_pred = ann.predict(X_test)

CPU times: user 12.8 ms, sys: 4.38 ms, total: 17.1 ms
Wall time: 3.17 ms


In [24]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1975
           1       0.47      0.24      0.32        71

    accuracy                           0.96      2046
   macro avg       0.72      0.61      0.65      2046
weighted avg       0.96      0.96      0.96      2046



## Neural Networks using Random Search CV

In [25]:
%%time

score_measure = "accuracy"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (50,), (70,),(50,30), (40,20), (60,40, 20), (70,50,40)],
    'activation': ['logistic', 'tanh', 'relu'],
    'solver': ['adam', 'sgd'],
    'alpha': [0, .2, .5, .7, 1],
    'learning_rate': ['constant', 'invscaling', 'adaptive'],
    'learning_rate_init': [0.001, 0.01, 0.1, 0.2, 0.5],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = RandomizedSearchCV(estimator = ann, param_distributions=param_grid, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

bestRecallTree = grid_search.best_estimator_

print(grid_search.best_params_)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-p

{'solver': 'adam', 'max_iter': 5000, 'learning_rate_init': 0.01, 'learning_rate': 'constant', 'hidden_layer_sizes': (60, 40, 20), 'alpha': 0.5, 'activation': 'tanh'}
CPU times: user 6.47 s, sys: 1.02 s, total: 7.49 s
Wall time: 2min 1s


In [26]:
%%time
y_pred = bestRecallTree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1975
           1       0.35      0.10      0.15        71

    accuracy                           0.96      2046
   macro avg       0.66      0.55      0.57      2046
weighted avg       0.95      0.96      0.95      2046

CPU times: user 46.5 ms, sys: 16.5 ms, total: 63 ms
Wall time: 13.3 ms


In [27]:
c_matrix = confusion_matrix(y_test, grid_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"NN_random", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## Neural Networks using Grid Search CV

In [28]:
%%time

score_measure = "accuracy"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (30,), (50,), (70,), (90,)],
    'activation': ['tanh', 'relu'],
    'solver': ['adam'],
    'alpha': [.5, .7, 1],
    'learning_rate': ['adaptive', 'invscaling'],
    'learning_rate_init': [0.005, 0.01, 0.15],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = GridSearchCV(estimator = ann, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

bestRecallTree = grid_search.best_estimator_

print(grid_search.best_params_)

Fitting 5 folds for each of 144 candidates, totalling 720 fits


/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/meghanaraju/opt/anaconda3/envs/tf/lib/python3.9/site-p

{'activation': 'tanh', 'alpha': 0.5, 'hidden_layer_sizes': (50,), 'learning_rate': 'invscaling', 'learning_rate_init': 0.005, 'max_iter': 5000, 'solver': 'adam'}
CPU times: user 5.03 s, sys: 1.15 s, total: 6.18 s
Wall time: 45.9 s


In [29]:
%%time
y_pred = bestRecallTree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1975
           1       0.41      0.10      0.16        71

    accuracy                           0.96      2046
   macro avg       0.69      0.55      0.57      2046
weighted avg       0.95      0.96      0.95      2046

CPU times: user 37.2 ms, sys: 11.6 ms, total: 48.8 ms
Wall time: 9.02 ms


In [30]:
c_matrix = confusion_matrix(y_test, grid_search.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"NN_grid", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

## DNN

In [9]:
import tensorflow as tf
from tensorflow import keras

# fix random seed for reproducibility
np.random.seed(1)
tf.random.set_seed(1)

In [10]:
X_train.shape

(4773, 22)

In [11]:
%%time

# create model stucture
model = keras.models.Sequential()
model.add(keras.layers.Input(shape=22))
model.add(keras.layers.Dense(22, activation='relu'))
model.add(keras.layers.Dense(22, activation='relu'))
model.add(keras.layers.Dense(22, activation='relu'))
model.add(keras.layers.Dense(1, activation='sigmoid')) # final layer, 10 categories


# compile
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# if you want to overide the defaults for the optimizer....
#adam = keras.optimizers.Adam(learning_rate=0.01)
#model.compile(loss='sparse_categorical_crossentropy', optimizer=adam, metrics=['accuracy'])

CPU times: user 34.6 ms, sys: 8.79 ms, total: 43.4 ms
Wall time: 71.3 ms


In [12]:
%%time

# fit the model
history = model.fit(X_train, y_train, 
                    validation_data=(X_test, y_test), 
                    epochs=20, batch_size=100)

Epoch 1/20
48/48 [==============================] - 0s 2ms/step - loss: 0.4764 - accuracy: 0.9409 - val_loss: 0.2962 - val_accuracy: 0.9653
Epoch 2/20
48/48 [==============================] - 0s 924us/step - loss: 0.2211 - accuracy: 0.9688 - val_loss: 0.1626 - val_accuracy: 0.9653
Epoch 3/20
48/48 [==============================] - 0s 880us/step - loss: 0.1505 - accuracy: 0.9688 - val_loss: 0.1317 - val_accuracy: 0.9653
Epoch 4/20
48/48 [==============================] - 0s 877us/step - loss: 0.1241 - accuracy: 0.9688 - val_loss: 0.1211 - val_accuracy: 0.9653
Epoch 5/20
48/48 [==============================] - 0s 893us/step - loss: 0.1096 - accuracy: 0.9690 - val_loss: 0.1157 - val_accuracy: 0.9658
Epoch 6/20
48/48 [==============================] - 0s 838us/step - loss: 0.1002 - accuracy: 0.9696 - val_loss: 0.1144 - val_accuracy: 0.9673
Epoch 7/20
48/48 [==============================] - 0s 880us/step - loss: 0.0947 - accuracy: 0.9715 - val_loss: 0.1138 - val_accuracy: 0.9673
Epoch 8/

In [13]:
# evaluate the model

scores = model.evaluate(X_test, y_test, verbose=0)
scores
# In results, first is loss, second is accuracy

[0.11455024778842926, 0.9652981162071228]

In [14]:
# let's format this into a better output...

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.11
accuracy: 96.53%


## Wide and Deep Networks

In [15]:
#Define the model: for multi-class

model = keras.models.Sequential()

model.add(keras.layers.Input(shape=22))
model.add(keras.layers.Dense(60, activation='relu'))
model.add(keras.layers.Dense(60, activation='relu'))
model.add(keras.layers.Dense(60, activation='relu'))
model.add(keras.layers.Dense(1, activation='sigmoid'))

In [16]:
# Compile model

#Optimizer:
adam = keras.optimizers.Adam(learning_rate=0.01)
model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])

In [17]:
# Fit the model

history = model.fit(X_train, y_train, 
                    validation_data=(X_test, y_test), 
                    epochs=20, batch_size=100)

Epoch 1/20
48/48 [==============================] - 0s 2ms/step - loss: 0.1743 - accuracy: 0.9625 - val_loss: 0.1176 - val_accuracy: 0.9653
Epoch 2/20
48/48 [==============================] - 0s 996us/step - loss: 0.0920 - accuracy: 0.9702 - val_loss: 0.1199 - val_accuracy: 0.9653
Epoch 3/20
48/48 [==============================] - 0s 951us/step - loss: 0.0846 - accuracy: 0.9713 - val_loss: 0.1118 - val_accuracy: 0.9653
Epoch 4/20
48/48 [==============================] - 0s 1ms/step - loss: 0.0800 - accuracy: 0.9698 - val_loss: 0.1179 - val_accuracy: 0.9653
Epoch 5/20
48/48 [==============================] - 0s 989us/step - loss: 0.0792 - accuracy: 0.9719 - val_loss: 0.1331 - val_accuracy: 0.9643
Epoch 6/20
48/48 [==============================] - 0s 986us/step - loss: 0.0757 - accuracy: 0.9709 - val_loss: 0.1226 - val_accuracy: 0.9648
Epoch 7/20
48/48 [==============================] - 0s 960us/step - loss: 0.0753 - accuracy: 0.9715 - val_loss: 0.1158 - val_accuracy: 0.9658
Epoch 8/20

In [18]:
# evaluate the model

scores = model.evaluate(X_test, y_test, verbose=0)
scores

# In results, first is loss, second is accuracy

[0.2160780131816864, 0.9623655676841736]

In [19]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.22
accuracy: 96.24%


## Random GridSearch using Glorot Uniform

In [20]:
%%time

# If you don't have the following installed, from command line '!pip install scikeras'
from scikeras.wrappers import KerasClassifier
from keras.initializers import GlorotNormal

score_measure = "accuracy"
kfolds = 5

def build_clf(hidden_layer_sizes, dropout):
    ann = tf.keras.models.Sequential()
    ann.add(keras.layers.Input(shape=22)),
    for hidden_layer_size in hidden_layer_sizes:
        model.add(keras.layers.Dense(hidden_layer_size, kernel_initializer= tf.keras.initializers.GlorotUniform(), 
                                     bias_initializer=keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None), activation="relu"))
        model.add(keras.layers.Dropout(dropout))
    ann.add(tf.keras.layers.Dense(1, activation='sigmoid'))
    ann.compile(loss = 'binary_crossentropy', metrics = ['accuracy'])
    return ann


CPU times: user 31 µs, sys: 2 µs, total: 33 µs
Wall time: 42.9 µs


In [21]:
from scikeras.wrappers import KerasClassifier

keras_clf = KerasClassifier(
    model=build_clf,
    hidden_layer_sizes=64,
    dropout = 0.0
)


In [22]:
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.model_selection import RandomizedSearchCV

params = {
    'optimizer__learning_rate': [0.0005, 0.001, 0.005],
    'model__hidden_layer_sizes': [(70,),(90, ), (100,), (100, 90)],
    'model__dropout': [0, 0.1],
    'batch_size':[20, 60, 100],
    'epochs':[10, 50, 100],
    'optimizer':["adam",'sgd']
}
keras_clf.get_params().keys()

dict_keys(['model', 'build_fn', 'warm_start', 'random_state', 'optimizer', 'loss', 'metrics', 'batch_size', 'validation_batch_size', 'verbose', 'callbacks', 'validation_split', 'shuffle', 'run_eagerly', 'epochs', 'hidden_layer_sizes', 'dropout', 'class_weight'])

In [23]:
rnd_search_cv = RandomizedSearchCV(estimator=keras_clf, param_distributions=params, scoring='accuracy', n_iter=50, cv=5)

import sys
sys.setrecursionlimit(10000) # note: the default is 3000 (python 3.9)

earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=0, mode='auto')
callback = [earlystop]

_ = rnd_search_cv.fit(X_train, y_train, callbacks=callback, verbose=0)

16/16 [==============================] - 0s 320us/step


In [24]:
rnd_search_cv.best_params_

{'optimizer__learning_rate': 0.0005,
 'optimizer': 'sgd',
 'model__hidden_layer_sizes': (100, 90),
 'model__dropout': 0.1,
 'epochs': 100,
 'batch_size': 60}

In [25]:
best_net = rnd_search_cv.best_estimator_
print(rnd_search_cv.best_params_)

{'optimizer__learning_rate': 0.0005, 'optimizer': 'sgd', 'model__hidden_layer_sizes': (100, 90), 'model__dropout': 0.1, 'epochs': 100, 'batch_size': 60}


In [26]:
%%time
y_pred = best_net.predict(X_test)
print(classification_report(y_test, y_pred))

35/35 [==============================] - 0s 333us/step
              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1975
           1       0.37      0.10      0.16        71

    accuracy                           0.96      2046
   macro avg       0.67      0.55      0.57      2046
weighted avg       0.95      0.96      0.95      2046

CPU times: user 49.5 ms, sys: 5.05 ms, total: 54.5 ms
Wall time: 47.3 ms


In [34]:
c_matrix = confusion_matrix(y_test, rnd_search_cv.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Glorot Uniform", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

35/35 [==============================] - 0s 451us/step


## Random Grid Search with Glorot Normal

In [35]:
%%time

# If you don't have the following installed, from command line '!pip install scikeras'
from scikeras.wrappers import KerasClassifier
from keras.initializers import GlorotNormal

score_measure = "accuracy"
kfolds = 5

def build_clf(hidden_layer_sizes, dropout):
    ann = tf.keras.models.Sequential()
    ann.add(keras.layers.Input(shape=22)),
    for hidden_layer_size in hidden_layer_sizes:
        model.add(keras.layers.Dense(hidden_layer_size, kernel_initializer= tf.keras.initializers.GlorotNormal(), 
                                     bias_initializer=keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None), activation="relu"))
        model.add(keras.layers.Dropout(dropout))
    ann.add(tf.keras.layers.Dense(1, activation='sigmoid'))
    ann.compile(loss = 'binary_crossentropy', metrics = ['accuracy'])
    return ann

CPU times: user 24 µs, sys: 59 µs, total: 83 µs
Wall time: 34.1 µs


In [ ]:
from scikeras.wrappers import KerasClassifier

keras_clf = KerasClassifier(
    model=build_clf,
    hidden_layer_sizes=64,
    dropout = 0.0
)

In [37]:
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.model_selection import RandomizedSearchCV

params = {
    'optimizer__learning_rate': [0.0005, 0.001, 0.005],
    'model__hidden_layer_sizes': [(70,),(90, ), (100,), (100, 90)],
    'model__dropout': [0, 0.1],
    'batch_size':[20, 60, 100],
    'epochs':[10, 50, 100],
    'optimizer':["adam",'sgd']
}
keras_clf.get_params().keys()

dict_keys(['model', 'build_fn', 'warm_start', 'random_state', 'optimizer', 'loss', 'metrics', 'batch_size', 'validation_batch_size', 'verbose', 'callbacks', 'validation_split', 'shuffle', 'run_eagerly', 'epochs', 'hidden_layer_sizes', 'dropout', 'class_weight'])

In [38]:
rnd_search_cv = RandomizedSearchCV(estimator=keras_clf, param_distributions=params, scoring='accuracy', n_iter=50, cv=5)

import sys
sys.setrecursionlimit(10000) # note: the default is 3000 (python 3.9)

earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=0, mode='auto')
callback = [earlystop]

_ = rnd_search_cv.fit(X_train, y_train, callbacks=callback, verbose=0)

16/16 [==============================] - 0s 318us/step


In [39]:
rnd_search_cv.best_params_

{'optimizer__learning_rate': 0.005,
 'optimizer': 'adam',
 'model__hidden_layer_sizes': (70,),
 'model__dropout': 0,
 'epochs': 100,
 'batch_size': 100}

In [40]:
best_net = rnd_search_cv.best_estimator_
print(rnd_search_cv.best_params_)

{'optimizer__learning_rate': 0.005, 'optimizer': 'adam', 'model__hidden_layer_sizes': (70,), 'model__dropout': 0, 'epochs': 100, 'batch_size': 100}


In [41]:
%%time
y_pred = best_net.predict(X_test)
print(classification_report(y_test, y_pred))

21/21 [==============================] - 0s 387us/step
              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1975
           1       0.38      0.08      0.14        71

    accuracy                           0.96      2046
   macro avg       0.67      0.54      0.56      2046
weighted avg       0.95      0.96      0.95      2046

CPU times: user 43.5 ms, sys: 4.21 ms, total: 47.7 ms
Wall time: 42.6 ms


In [42]:
c_matrix = confusion_matrix(y_test, rnd_search_cv.predict(X_test))
TP = c_matrix[1][1]
TN = c_matrix[0][0]
FP = c_matrix[0][1]
FN = c_matrix[1][0]
performance = pd.concat([performance, pd.DataFrame({'model':"Glorot Normal", 
                                                    'Accuracy': [(TP+TN)/(TP+TN+FP+FN)], 
                                                    'Precision': [TP/(TP+FP)], 
                                                    'Recall': [TP/(TP+FN)], 
                                                    'F1': [2*TP/(2*TP+FP+FN)]
                                                     }, index=[0])])

21/21 [==============================] - 0s 346us/step


In [31]:
performance         # Old Results

model  Accuracy  Precision    Recall        F1
0  Logistic RandomRegression  0.964321   0.428571  0.084507  0.141176
0    Logistic GridRegression  0.964321   0.428571  0.084507  0.141176
0                 SVM Random  0.964809   0.444444  0.056338  0.100000
0                   SVM Grid  0.964809   0.444444  0.056338  0.100000
0               Dtree_random  0.965298        NaN  0.000000  0.000000
0                 Dtree_grid  0.965298        NaN  0.000000  0.000000
0                  NN_random  0.962366   0.350000  0.098592  0.153846
0                    NN_grid  0.963832   0.411765  0.098592  0.159091

In [43]:
performance         # New Results

model  Accuracy  Precision    Recall        F1
0  Glorot Uniform  0.962854   0.368421  0.098592  0.155556
0   Glorot Normal  0.963343   0.375000  0.084507  0.137931

## Analysis - MLP Classifier and Keras model

Firstly, the above table shows results for 10 different models: Logistic RandomRegression, Logistic GridRegression, SVM Random, SVM Grid, Dtree random, Dtree grid, NN_random, NN_grid, Glorot Uniform and Glorot Normal. 

Here our main focus is on Neural Networks using MLP Classifier and Deep Neural netowrk using "Glort Normal and Glorot Uniform".

And also we are considering, Loss as "Binary" and Activation as "Sigmoid" function as the target varible is "Binary Classifier". 

And all Models : accuracy, precision, recall, and F1 score were analyzed. Here, we are focusing on the 'accuracy' metrics. 

By considering the NN_random, NN_grid, Glorot Uniform and Glorot Normal. The NN_grid model obtained the highest accuracy score of "0.963832" and had the highest precision score of "0.411765" indicating that it was the most effective at identifying true positives.